platform/mellanox/mlnx-platform-api/sonic_platform/chassis.py

```diff
@@ -95,5 +129,169 @@ def _get_port_position_tuple_by_sku_name(self):
+        """
+        Read a generic file, returns the contents of the file
+        """
+        r = ''
+        try:
+            f = io.open(filename)
+            r = f.read(len)
+            f.close()
+            return r
+        except:
+            return ''
+
+    def _verify_reboot_cause(self, filename):
+        '''
+        Open and read the reboot cause file in 
+        /var/run/hwmanagement/system (which is defined as REBOOT_CAUSE_ROOT)
+        If a reboot cause file doesn't exists, returns '0'.
+        '''
+        return bool(int(self._read_generic_file(join(REBOOT_CAUSE_ROOT, filename), REBOOT_CAUSE_FILE_LENGTH).rstrip('\n')))
+
+    def get_reboot_cause(self):
+        """
+        Retrieves the cause of the previous reboot
+
+        Returns:
+            A tuple (string, string) where the first element is a string
+            containing the cause of the previous reboot. This string must be
+            one of the predefined strings in this class. If the first string
+            is "REBOOT_CAUSE_HARDWARE_OTHER", the second string can be used
+            to pass a description of the reboot cause.
+        """
+        #read reboot causes files in the following order
+        minor_cause = ''
+        if self._verify_reboot_cause(REBOOT_CAUSE_POWER_LOSS_FILE):
+            major_cause = self.REBOOT_CAUSE_POWER_LOSS
+        elif self._verify_reboot_cause(REBOOT_CAUSE_THERMAL_OVERLOAD_ASIC_FILE):
+            major_cause = self.REBOOT_CAUSE_THERMAL_OVERLOAD_ASIC
+        elif self._verify_reboot_cause(REBOOT_CAUSE_WATCHDOG_FILE):
+            major_cause = self.REBOOT_CAUSE_WATCHDOG
+        else:
+            major_cause = self.REBOOT_CAUSE_HARDWARE_OTHER
+            if self._verify_reboot_cause(REBOOT_CAUSE_MLNX_FIRMWARE_RESET):
+                minor_cause = "Reset by ASIC firmware"
+            else:
+                reboot_by_user = self._read_generic_file(PREVIOUS_REBOOT_CAUSE_FILE, PREVIOUS_REBOOT_CAUSE_MAX_LENGTH)
+                if '' != reboot_by_user:
+                    major_cause = self.REBOOT_CAUSE_NON_HARDWARE
+                    minor_cause = reboot_by_user
+
+        return major_cause, minor_cause
+
+    def _get_cpld_version(self):
+        cpld1_version = self._read_generic_file(join(CPLD_VERSION_ROOT, CPLD1_VERSION_FILE), CPLD_VERSION_MAX_LENGTH)
+        cpld2_version = self._read_generic_file(join(CPLD_VERSION_ROOT, CPLD2_VERSION_FILE), CPLD_VERSION_MAX_LENGTH)
+        return "cpld1 version: " + cpld1_version.rstrip('\n') + " cpld 2 version: " + cpld2_version.rstrip('\n')
+
+    def _get_command_result(self, cmdline):
+        try:
+            proc = subprocess.Popen(cmdline, stdout=subprocess.PIPE, shell=True, stderr=subprocess.STDOUT)
+            stdout = proc.communicate()[0]
+            proc.wait()
+            result = stdout.rstrip('\n')
+
+        except OSError, e:
+            result = ''
+
+        return result
+
+    def _get_firmware_version(self):
```

Comment:
What is referred as FW in the chassis? FW ASIC? not sure. FW can also be referred to ONIE version. Just want to be sure we are aligned with requirement

Reply:
In my opinion, this API is designed to retrieve the version of some core components on a switch in an as-much-as-possible way so that the client/administrator can make some decision based on it. For example, whether or not to deploy a workaround for an issue related to a specific version on a switch can be easily determined by checking the version of it.
Anyway, I will check with Joe and Kebo.
@jleveque @keboliu, could you please confirm that?

Reply:
Yes Stephen, the intended purpose of get_firmware_version() is to retrieve the current firmware version of a component on the chassis (BIOS, ASIC, FPGA, etc.) to compare and determine if an update is necessary.